go/vt/mysqlctl/backup.go

```diff
@@ -0,0 +1,548 @@
+			dbName := row[0].String()
+			tableQr, err := mysqld.FetchSuperQuery("SHOW TABLES FROM " + dbName)
+			if err != nil {
+				return fmt.Errorf("checkNoDB failed: %v", err)
+			} else if len(tableQr.Rows) == 0 {
+				// no tables == empty db, all is well
+				continue
+			}
+			return fmt.Errorf("checkNoDB failed, found active db %v", dbName)
+		}
+	}
+
+	return nil
+}
+
+// restoreFiles will copy all the files from the BackupStorage to the
+// right place
+func restoreFiles(cnf *Mycnf, bh backupstorage.BackupHandle, fes []FileEntry, restoreConcurrency int) error {
+	sema := sync2.NewSemaphore(restoreConcurrency, 0)
+	rec := concurrency.AllErrorRecorder{}
+	wg := sync.WaitGroup{}
+	for i, fe := range fes {
+		wg.Add(1)
+		go func(i int, fe FileEntry) {
+			defer wg.Done()
+
+			// wait until we are ready to go, skip if we already
+			// encountered an error
+			sema.Acquire()
+			defer sema.Release()
+			if rec.HasErrors() {
+				return
+			}
+
+			// open the source file for reading
+			name := fmt.Sprintf("%v", i)
+			source, err := bh.ReadFile(name)
+			if err != nil {
+				rec.RecordError(err)
+				return
+			}
+			defer source.Close()
+
+			// open the destination file for writing
+			dstFile, err := fe.open(cnf, false)
+			if err != nil {
+				rec.RecordError(err)
+				return
+			}
+			defer dstFile.Close()
+
+			// create a buffering output
+			dst := bufio.NewWriterSize(dstFile, 2*1024*1024)
+
+			// create hash to write the compressed data to
+			hasher := newHasher()
+
+			// create a Tee: we split the input into the hasher
+			// and into the gunziper
+			tee := io.TeeReader(source, hasher)
+
+			// create the uncompresser
+			gz, err := cgzip.NewReader(tee)
+			if err != nil {
+				rec.RecordError(err)
+				return
+			}
+			defer gz.Close()
+
+			// copy the data. Will also write to the hasher
+			if _, err = io.Copy(dst, gz); err != nil {
+				rec.RecordError(err)
+				return
+			}
+
+			// check the hash
+			hash := hasher.HashString()
+			if hash != fe.Hash {
+				rec.RecordError(fmt.Errorf("hash mismatch for %v, got %v expected %v", fe.Name, hash, fe.Hash))
+				return
+			}
+
+			// flush the buffer
+			dst.Flush()
+		}(i, fe)
+	}
+	wg.Wait()
+	return rec.Error()
+}
+
+// Restore is the main entry point for backup restore.  If there is no
+// appropriate backup on the BackupStorage, Restore logs an error
+// and returns ErrNoBackup. Any other error is returned.
+func Restore(mysqld MysqlDaemon, bucket string, restoreConcurrency int, hookExtraEnv map[string]string) (proto.ReplicationPosition, error) {
+	// find the right backup handle: most recent one, with a MANIFEST
+	log.Infof("Restore: looking for a suitable backup to restore")
+	bs := backupstorage.GetBackupStorage()
+	bhs, err := bs.ListBackups(bucket)
+	if err != nil {
+		return proto.ReplicationPosition{}, fmt.Errorf("ListBackups failed: %v", err)
+	}
+	toRestore := len(bhs) - 1
+	var bh backupstorage.BackupHandle
+	var bm BackupManifest
+	for toRestore >= 0 {
```

Comment:
Added a comment to say ListBackup returns them ordered. The other use case is 'vtctl ListBackups' which returns them in ascending order (old to new) so I think this is good. I need the loop anyway to find the latest correct one.